File: app/controllers/conditions_controller.rb
# frozen_string_literal: true

require 'alerts/configuration'

class ConditionsController < ApplicationController
  before_action :require_user!
  before_action :set_alerts_configuration

  def fields
    @index = params[:index]
    @provider_type = params[:provider_type]
    @source = params[:source]
    @field = @alerts_configuration.source_fields(@provider_type, @source).first

    respond_to do |format|
      format.turbo_stream
    end
  end

  def value
    @index = params[:index]
    @provider_type = params[:provider_type]
    @source = params[:source]
    @field = params[:field]

    respond_to do |format|
      format.turbo_stream
    end
  end

  def add
    @index = params[:index].to_i
    @alert =
      if params[:alert_id]
        current_user.alerts.find(params[:alert_id])
      else
        current_user.alerts.new
      end
    @provider_type = params[:provider_type]
    @condition = @alert.conditions.build

    respond_to do |format|
      format.turbo_stream
    end
  end

  def delete
    @index = params[:index].to_i
    @alert = current_user.alerts.find(params[:alert_id])
    @condition = @alert.conditions[@index]

    respond_to do |format|
      format.turbo_stream
    end
  end

  private

  def set_alerts_configuration
    @alerts_configuration = Alerts::Configuration.new
  end
end

File: app/javascript/controllers/conditions_controller.js
import { Controller } from "@hotwired/stimulus";
import { get } from "@rails/request.js";

// Connects to data-controller="conditions"
export default class extends Controller {
  changeSource(event) {
    const index = event.target.id.split("_")[3]; // alert_conditions_attributes_1_source
    const providerType = document.getElementById("alert_provider_type").value;
    const source = event.target.selectedOptions[0].value;

    get(
      `/conditions/fields?index=${index}&provider_type=${providerType}&source=${source}`,
      {
        responseKind: "turbo-stream",
      }
    );
  }

  changeField(event) {
    const index = event.target.id.split("_")[3]; // alert_conditions_attributes_1_field
    const providerType = document.getElementById("alert_provider_type").value;
    const source = document.getElementById(
      `alert_conditions_attributes_${index}_source`
    ).selectedOptions[0].value;
    const field = event.target.selectedOptions[0].value;

    get(
      `/conditions/value?index=${index}&provider_type=${providerType}&source=${source}&field=${field}`,
      {
        responseKind: "turbo-stream",
      }
    );
  }

  add(event) {
    event.preventDefault();

    const index = document.getElementsByClassName("condition").length; // already +1 for zero based index
    const formUrl = new URL(event.target.form.action);
    let addUrl;
    if (formUrl.pathname === "/alerts") {
      addUrl = `/conditions/add?index=${index}`;
    } else {
      const alertId = formUrl.pathname.split("/")[2]; // /alerts/1
      addUrl = `/conditions/add?index=${index}&alert_id=${alertId}`;
    }

    get(addUrl, {
      responseKind: "turbo-stream",
    });
  }

  delete(event) {
    event.preventDefault();

    const index = event.currentTarget.id.split("_")[2]; // alert_conditions_1_destroy_button
    const $destroyField = document.getElementById(
      `alert_conditions_attributes_${index}__destroy`
    );

    if ($destroyField) {
      // if persisted
      // replace with strikethrough condition string
      // and set _destroy attribute to true via turbo
      const formUrl = new URL(event.currentTarget.form.action);
      const alertId = formUrl.pathname.split("/")[2]; // /alerts/1
      const deleteUrl = `/conditions/delete?index=${index}&alert_id=${alertId}`;

      get(deleteUrl, {
        responseKind: "turbo-stream",
      });
    } else {
      // remove the two condition elements
      const $condition = event.currentTarget.parentElement.parentElement;
      const $conditionLabel = $condition.previousElementSibling;

      $condition.remove();
      $conditionLabel.remove();
    }
  }
}
